core/trino-main/src/main/java/io/trino/sql/gen/columnar/BloomFilter.java

```diff
@@ -0,0 +1,192 @@
+            if (block.mayHaveNull()) {
+                for (int position = offset; position < offset + size; position++) {
+                    boolean result = block.isNull(position) ? isNullAllowed : filter.test(block, position);
+                    outputPositions[selectedPositionsCount] = position;
+                    selectedPositionsCount += result ? 1 : 0;
+                }
+                return selectedPositionsCount;
+            }
+
+            for (int position = offset; position < offset + size; position++) {
+                outputPositions[selectedPositionsCount] = position;
+                selectedPositionsCount += filter.test(block, position) ? 1 : 0;
+            }
+            return selectedPositionsCount;
+        }
+
+        @Override
+        public int filterPositionsList(ConnectorSession session, int[] outputPositions, int[] activePositions, int offset, int size, Page page)
+        {
+            VariableWidthBlock block = (VariableWidthBlock) page.getBlock(0);
+            int selectedPositionsCount = 0;
+            if (block.mayHaveNull()) {
+                for (int index = offset; index < offset + size; index++) {
+                    int position = activePositions[index];
+                    boolean result = block.isNull(position) ? isNullAllowed : filter.test(block, position);
+                    outputPositions[selectedPositionsCount] = position;
+                    selectedPositionsCount += result ? 1 : 0;
+                }
+                return selectedPositionsCount;
+            }
+
+            for (int index = offset; index < offset + size; index++) {
+                int position = activePositions[index];
+                outputPositions[selectedPositionsCount] = position;
+                selectedPositionsCount += filter.test(block, position) ? 1 : 0;
+            }
+            return selectedPositionsCount;
+        }
+
+        @Override
+        public InputChannels getInputChannels()
+        {
+            return inputChannels;
+        }
+    }
+
+    public static final class SliceBloomFilter
+    {
+        private final long[] bloom;
+        private final int bloomSizeMask;
+
+        /**
+         * A Bloom filter for a set of Slice values.
+         * This is approx 2X faster than the Bloom filter implementations in ORC and parquet because
+         * it uses single hash function and uses that to set 3 bits within a 64 bit word.
+         * The memory footprint is up to (4 * values.size()) bytes, which is much smaller than maintaining a hash set of strings.
+         *
+         * @param values List of values used for filtering
+         */
+        public SliceBloomFilter(List<Slice> values)
+        {
+            int bloomSize = getBloomFilterSize(values.size());
+            bloom = new long[bloomSize];
+            bloomSizeMask = bloomSize - 1;
+            for (Slice value : values) {
+                long hashCode = XxHash64.hash(value);
```

Comment:
Slice has a hashCode that is using XxHash64 already (and is memoized). Just `value.hashCode()`

Reply:
These Slices are temporary objects that are created from a single contiguous Block, depending on the Type the Slice may be subject to truncation and padding as well.
So I don't think we gain anything by memoized hash code.
On the other hand, the hashing logic for bloom filter could evolve to be different from Slice's hashCode implementation.